### PR TITLE
Mongo Collation Support

### PR DESCRIPTION
**Pull request by mvogttech**
- Add full MongoDB [collation](https://www.mongodb.com/docs/manual/reference/collation/) support to `find`, `findOne`, `findOneAsync`, and `observeChanges`
- Implement collation-aware string comparison in Minimongo using `Intl.Collator`, so client and server behave identically
- Compatible with oplog-tailing — no fallback to polling required

## Motivation

Previously, collation options passed to `Collection.find()` were silently dropped by the mongo driver's cursor whitelist, and Minimongo had no concept of locale-aware string comparison at all. This forced workarounds like `generateCasePermutationsForString`, which builds 16-clause `$or` queries for case-insensitive lookups.
> **Note:** This is how I found minimongo/mongo were missing collation support. I wanted to modernize accounts-* packages and ran into `generateCasePermutationsForString` -- which is a workaround that exists since MongoDB v2.

With this PR, a simple `{collation: {locale: 'en', strength: 2}}` option enables case-insensitive matching on both client and server, using MongoDB's native collation on the server and `Intl.Collator` in Minimongo. Since Minimongo now understands collation, the oplog observe driver can correctly filter oplog events for collation queries without falling back to polling.

### Minimongo (`packages/minimongo/`)

| File | Change |
|------|--------|
| `matcher.js` | `Matcher` accepts `collation` param, creates `Intl.Collator` via `_createCollator`. Threads collator through `_cmp` and `_equal` for all string comparisons. Scalar `_id` fast path falls through to document selector compilation when collation is active. |
| `common.js` | `equalityElementMatcher` and inequality operators (`$gt`, `$gte`, `$lt`, `$lte`, `$eq`, `$ne`, `$in`, `$nin`) accept and pass through the collator. |
| `sorter.js` | `Sorter` accepts `collation` param, passes collator to `_cmp` in `_keyFieldComparator`. |
| `cursor.js` | Passes `options.collation` to `Matcher` and `Sorter`. Skips `_selectorId` fast path when collation is present (direct `IdMap.get` is not collation-aware). |

### Mongo driver (`packages/mongo/`)

| File | Change |
|------|--------|
| `mongo_connection.js` | Added `collation` to the `mongoOptions` whitelist in `_createAsynchronousCursor`. Passes `cursorDescription.options.collation` to `Minimongo.Matcher` and `Minimongo.Sorter` in `_observeChanges`. |
| `cursor_description.ts` | Added `CollationOptions` interface and `collation?` field to `CursorOptions`. |
| `package.js` | Registered `collation_tests.js` in `onTest`. |

### Documentation

| File | Change |
|------|--------|
| `collection/methods_sync.js` | Added `@param options.collation` to `find()` and `findOne()` JSDoc. |
| `collection/methods_async.js` | Added `@param options.collation` to `findOneAsync()` JSDoc. |
| `docs/source/api/collections.md` | Added paragraph explaining collation support with usage example. |
| `packages/minimongo/README.md` | Added collation to the feature list. |

## Collation option mapping

MongoDB collation options are mapped to `Intl.Collator` as follows:

| MongoDB | Intl.Collator | Effect |
|---------|--------------|--------|
| `strength: 1` | `sensitivity: 'base'` | Ignore case and accents (`cafe` = `Café`) |
| `strength: 2` | `sensitivity: 'accent'` | Case-insensitive (`alice` = `Alice`, `cafe` ≠ `café`) |
| `strength: 3` (default) | `sensitivity: 'variant'` | Exact match (default behavior) |
| `strength: 1` + `caseLevel: true` | `sensitivity: 'case'` | Case-sensitive but accent-insensitive |
| `numericOrdering: true` | `numeric: true` | `'2'` sorts before `'10'` |
| `caseFirst: 'upper'/'lower'` | `caseFirst: 'upper'/'lower'` | Uppercase or lowercase first in sort |

## Usage example

```js
// Case-insensitive find
const users = Users.find(
  { email: 'Alice@Example.COM' },
  { collation: { locale: 'en', strength: 2 } }
).fetch();

// Case-insensitive sort
const sorted = Posts.find(
  {},
  { collation: { locale: 'en', strength: 2 }, sort: { title: 1 } }
).fetch();

// Create a collation-backed index for efficient queries
await Users.createIndexAsync(
  { email: 1 },
  { collation: { locale: 'en', strength: 2 } }
);
```

### CODE REVIEW

File: packages/mongo/tests/collation_tests.js [hunk 1]
```diff
@@ -0,0 +1,314 @@
+// Tests for MongoDB collation support in the mongo package.
+// These test that collation options flow through to the MongoDB driver
+// and that observeChanges works correctly with collation-aware Minimongo.
+
+var makeCollection = function () {
+  return new Mongo.Collection('collation_' + Random.id());
+};
+
+if (Meteor.isServer) {
```
Comment (radekmie) [inline packages/mongo/tests/collation_tests.js:169]: Can some of these tests be run on the client as well? This way, we'd ensure compatibility between the implementations.

Comment (mvogttech) [reply packages/mongo/tests/collation_tests.js:169]: The client-side collation behavior is already covered by 9 tests in packages/minimongo/minimongo_tests_client.js - they test the Matcher, Sorter, LocalCollection.find, strength levels, numericOrdering, and _id matching with collation. 

Since Mongo.Collection(null) on the client just wraps LocalCollection, duplicating those here wouldn't add coverage, right? The tests in this file are intentionally server-only because they validate the full integration path through the MongoDB driver (createIndex, oplog, countDocuments, etc.).

What did you have in mind?

**Comment by mvogttech** [comment]
Also related: https://forums.meteor.com/t/slow-mongo-query-on-createuser/54879

File: packages/mongo/tests/collation_tests.js (hunk 1 above)
Comment (radekmie) [reply packages/mongo/tests/collation_tests.js:169]: I thought it'd be better to have a few tests run on both sides to make sure the same code results in the same behavior without relying on people keeping two files in sync, that's it.

Comment (mvogttech) [reply packages/mongo/tests/collation_tests.js:169]: Got it! I'll work on those.

**Comment by mvogttech** [comment]
@italojs - Do you want me to resolve the conflicts?

**Comment by italojs** [comment]
@mvogttech feel free to handle it. thanks ;)

File: packages/mongo/cursor_description.ts [hunk 2]
```diff
@@ -1,9 +1,21 @@
+interface CollationOptions {
+  locale: string;
+  caseLevel?: boolean;
+  caseFirst?: 'upper' | 'lower' | 'off';
+  strength?: 1 | 2 | 3 | 4 | 5;
+  numericOrdering?: boolean;
+  alternate?: 'non-ignorable' | 'shifted';
+  maxVariable?: 'punct' | 'space';
+  backwards?: boolean;
```
Comment (nachocodoner) [inline packages/mongo/cursor_description.ts:25]: We are exposing several collation options in the types here, but I'm not sure Minimongo actually implements all of them. For example, do alternate, maxVariable, backwards, and the other MongoDB collation options also have client-side support, or are they server-only for now?

As we don't have client-server-aware types, we may need to mention in the docs which of these are server-only.

File: packages/mongo/tests/collation_tests.js [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+      // Insert another doc that matches via collation
+      await c.insertAsync({ _id: 'z', name: 'ALICE' });
+      await new Promise(resolve => setTimeout(resolve, 200));
+      test.equal(addedDocs.length, 2);
+      test.equal(addedDocs[1].id, 'z');
+      test.equal(addedDocs[1].fields.name, 'ALICE');
+
+      // Remove original Alice — should trigger removed
+      await c.removeAsync('x');
+      await new Promise(resolve => setTimeout(resolve, 200));
```
Comment (nachocodoner) [inline packages/mongo/tests/collation_tests.js]: Instead of using setTimeout here, why not use [the waitUntil helper](https://github.com/meteor/meteor/blob/devel/packages/test-helpers/wait.js?rgh-link-date=2026-03-20T07%3A42%3A17Z#L5-L50) to enforce behavior-based waiting?

This will help on flakiness tests.

Example:

```
// Remove original Alice — should trigger removed
await c.removeAsync('x');
await waitUntil(
  () => removedIds.length === 1,
  { description: 'observer received removed callback after removing original Alice' }
);
test.equal(removedIds[0], 'x');
```

File: packages/mongo/cursor_description.ts (hunk 2 above)
Comment (mvogttech) [reply packages/mongo/cursor_description.ts:25]: You're right. Minimongo doesn't implement all collation options client-side. Here's the breakdown:

 Supported client-side (mapped to `Intl.Collator` in `packages/minimongo/matcher.js`):
  - `locale`
  - `strength` (1–3 only)
  - `caseLevel`
  - `numericOrdering`
  - `caseFirst`

Server-only (silently ignored by Minimongo):
  - `alternate`
  - `maxVariable`
  - `backwards`
  - `strength` values 4 & 5 (MongoDB-specific, no `Intl.Collator` equivalent)

I'll add JSDoc comments and update the docs to the `CollationOptions` fields in `cursor_description.ts` noting which ones are server-only.

File: packages/mongo/tests/collation_tests.js (hunk 3 above)
Comment (mvogttech) [reply packages/mongo/tests/collation_tests.js]: Good call. `waitUntil` is definitely better here. I'll update all three `setTimeout` waits in this test.

File: packages/minimongo/matcher.js [hunk 4]
```diff
@@ -49,6 +49,12 @@ export default class Matcher {
     // translated into {_id: ID} first. Used by canBecomeTrueByModifier and
     // Sorter._useWithMatcher.
     this._selector = null;
+    // An optional Intl.Collator for locale-aware string comparison (mirrors
+    // MongoDB's collation option).  Stored as the collator instance so that
+    // _cmp / _equal can use it without re-creating it on every call.
+    this._collation = collation
+      ? LocalCollection._createCollator(collation)
+      : null;
```
Comment (radekmie) [inline packages/minimongo/matcher.js]: Both `Matcher` and `Sorter` do it - it would be better to do it in `Cursors` and only pass down.

File: packages/minimongo/matcher.js [hunk 5]
```diff
@@ -357,3 +380,29 @@ LocalCollection._f = {
...
+    }
+  }
+  if (collation.numericOrdering != null) {
+    options.numeric = collation.numericOrdering;
+  }
+  if (collation.caseFirst != null && collation.caseFirst !== 'off') {
+    options.caseFirst = collation.caseFirst;
+  }
+  return new Intl.Collator(collation.locale, options);
+};
```
Comment (radekmie) [inline packages/minimongo/matcher.js:431]: 1. I'd rather avoid object creation in a rather hot path (line 398).
2. How about a proper validation of the `collation` object in development?

**Review by radekmie — COMMENTED** [review]
It's hard for me to tell, whether `Intl.Collator` and MongoDB's collation are truly compatible, but overall it looks correct. I'm only thinking about making some of the tests run both on the client and the server to check for that.

File: packages/minimongo/matcher.js (hunk 4 above)
Comment (mvogttech) [reply packages/minimongo/matcher.js]: Good point. I'll refactor so Cursor creates the `Intl.Collator` once and passes the instance to both Matcher and Sorter, instead of having each create their own. Matcher and Sorter will accept either a collation spec (for standalone use) or a pre-built collator.

File: packages/minimongo/matcher.js (hunk 5 above)
Comment (mvogttech) [reply packages/minimongo/matcher.js:431]: 1. You're right, I'll replace the inline object literal with an `if/else` chain (or a module-level constant map) so we're not allocating on every call.
2. Makes sense. I'll add a development-only validation that checks `locale` is a non-empty string, `strength` is 1–5 (with a warning for 4/5 on the client), and the other options are the expected types. Wrapped in `Meteor.isDevelopment` check so it's zero-cost in production.

**Comment by mvogttech** [comment]
> It's hard for me to tell, whether `Intl.Collator` and MongoDB's collation are truly compatible, but overall it looks correct. I'm only thinking about making some of the tests run both on the client and the server to check for that.

The client+server test parity should help catch any Intl.Collator vs MongoDB divergence — both sides now run the same find, sort, inequality, and strength-1 tests against the same data. If we find edge cases where they diverge, we can add targeted tests for those.

Based on my research Mongo's collation is based on Int.Collator. They are very similar.

File: packages/minimongo/matcher.js [hunk 6]
```diff
@@ -50,11 +50,13 @@ export default class Matcher {
...
-    this._collation = collation
-      ? LocalCollection._createCollator(collation)
-      : null;
+    // MongoDB's collation option).  Accepts either a pre-built Intl.Collator
+    // (when created by Cursor) or a MongoDB collation spec object.
+    this._collation = !collation
+      ? null
+      : collation instanceof Intl.Collator
+        ? collation
+        : LocalCollection._createCollator(collation);
```
Comment (radekmie) [inline packages/minimongo/matcher.js:59]: 1. I understand that you want to accept both. In that case, we could make `_createCollator` handle this logic, so it won't be duplicated. In that case, we can also make it handle `null`, too.
2. The property should be called `_collator`, not `_collation`.

**Review by radekmie — APPROVED** [review]
As far as I'm concerned, this one looks solid.
